src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java

```diff
@@ -0,0 +1,146 @@
+
+    private final ObjectMapper mapper = JsonMapper.builder().build();
+    private final Map<SchemaSpec, Validator> validators;
+    private final URI testSchemaUri = URI.create("urn:test");
+    private Map<URI, String> remotes = Map.of();
+
+    public DevHarrelImplementation() {
+        final SchemaResolver schemaResolver =
+                uri -> {
+                    final String resolved = remotes.get(URI.create(uri));
+                    if (resolved == null) {
+                        return SchemaResolver.Result.empty();
+                    }
+                    return SchemaResolver.Result.fromString(resolved);
+                };
+        final Validator validator2020 =
+                new dev.harrel.jsonschema.ValidatorFactory()
+                        .withSchemaResolver(schemaResolver)
+                        .createValidator();
+        final Validator validator2019 =
+                new dev.harrel.jsonschema.ValidatorFactory()
+                        .withDialect(new Dialects.Draft2019Dialect())
+                        .withSchemaResolver(schemaResolver)
+                        .createValidator();
+        /* Validate against meta-schemas in order to parse them eagerly */
+        validator2020.validate(URI.create(SpecificationVersion.DRAFT2020_12.getId()), "{}");
+        validator2019.validate(URI.create(SpecificationVersion.DRAFT2019_09.getId()), "{}");
+
+        validator2020.registerSchema(testSchemaUri, TestSchemas.DRAFT_2020_SCHEMA);
+
+        this.validators =
+                Map.of(
+                        DRAFT_2020_12, validator2020,
+                        DRAFT_2019_09, validator2019);
+    }
+
+    @Override
+    public MetaData metadata() {
+        return METADATA;
+    }
+
+    @Override
+    public JsonValidator prepare(
+            final String schema, final SchemaSpec spec, final AdditionalSchemas additionalSchemas) {
+        DevHarrelImplementation.this.remotes = additionalSchemas.remotes();
+        final Validator validator = validators.get(spec);
+        final URI schemaUri = validator.registerSchema(schema);
+
+        return new JsonValidator() {
+            @Override
+            public void validate(final String json) {
+                final Validator.Result result = validator.validate(schemaUri, json);
+                if (!result.isValid()) {
+                    throw new RuntimeException();
+                }
+            }
+
+            @Override
+            public byte[] serialize(final TestModel model, final boolean validate) {
+                try {
+                    final String asString = mapper.writeValueAsString(model);
+                    if (validate
+                            && !validators
+                                    .get(DRAFT_2020_12)
+                                    .validate(testSchemaUri, asString)
+                                    .isValid()) {
+                        throw new RuntimeException();
```

Comment:
Does your library expose details of _why_ it failed?  

Can be useful to expose these in the exception, as it is logged...

Reply:
If I understand correctly I am required to throw if the validation fails, right? Is it being logged even if it's expected to fail?

Reply:
> If I understand correctly I am required to throw if the validation fails, right?

Correct.

> Is it being logged even if it's expected to fail?

I've not checked the code, but I wouldn't of thought so. It should only be logged if the test was supposed to pass.  Exposing the error message can make it easier to track down the cause of false negatives.

Reply:
Hmm, I just don't want to add more overhead to the validation logic. Gonna try to make as lightweight as it can be